core/node/consensus/src/storage/denormalizer.rs

```diff
@@ -0,0 +1,127 @@
+use anyhow::Context;
+use zksync_concurrency::{ctx, ctx::Ctx, scope, time};
+use zksync_consensus_roles::{attester, validator};
+
+use crate::storage::{vm_reader::VMReader, CommitteeAttester, CommitteeValidator, ConnectionPool};
+
+#[derive(Debug)]
+pub struct Denormalizer {
+    pool: ConnectionPool,
+    reader: VMReader,
+}
+
+impl Denormalizer {
+    pub fn new(pool: ConnectionPool, reader: VMReader) -> Denormalizer {
+        Self { pool, reader }
+    }
+
+    pub async fn run(mut self, ctx: &Ctx) -> anyhow::Result<()> {
+        tracing::info!("Running zksync_node_consensus::storage::Denormalizer");
+        let res = scope::run!(ctx, |ctx, s| async {
+            let mut next_batch = self.last_batch_committees(ctx).await?.next();
+            loop {
+                self.wait_batch(ctx, next_batch).await?;
+                self.denormalize_batch_committees(ctx, next_batch).await?;
+                next_batch = next_batch.next();
+            }
+        })
+        .await;
+        match res {
+            Ok(()) | Err(ctx::Error::Canceled(_)) => Ok(()),
+            Err(ctx::Error::Internal(err)) => Err(err),
+        }
+    }
+
+    async fn denormalize_batch_committees(
```

Comment:
by "denormalizer" you mean database denormalization? I think it is a way too big a mental shortcut. How about sth more like "CommitteeStateExtractor", or sth else which would actually indicate that this is the thing extracting the committee from the L2 state at the end of the batch.

Reply:
I used [database] “denormalizer” to describe an ongoing read/rewrite procedure. I didn’t add (yet) a noun to describe the domain, just to keep extendable. 

As for the verb, I think “extractor” is even more abstract. 
It can make sense if it’s about finding/setting the committees to be used, however I see it as a mere database query optimization / copying of data from one place to another. 

If you don’t want to go for database lingo, I can change that.

Reply:
I mean, it is the schema of the db that can be denormalized, not the data itself (which is just duplicated in storage then).

Reply:
Changed to `CommitteeExtractor`.